src/app/ReadClient.cpp

```diff
@@ -517,6 +541,118 @@ CHIP_ERROR ReadClient::ProcessAttributeReportIBs(TLV::TLVReader & aAttributeRepo
+        hasDeltaSystemTimestamp             = true;
+    }
+
+    err = aEventData.GetDeltaEpochTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mDeltaTimestamp.mType  = Timestamp::Type::kEpoch;
+        aEventHeader.mDeltaTimestamp.mValue = timeStampVal;
+        hasDeltaEpochTimestamp              = true;
+    }
+
+    err = aEventData.GetSystemTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mTimestamp.mType  = Timestamp::Type::kSystem;
+        aEventHeader.mTimestamp.mValue = timeStampVal;
+        hasSystemTimestamp             = true;
+    }
+
+    err = aEventData.GetEpochTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mDeltaTimestamp.mType  = Timestamp::Type::kEpoch;
+        aEventHeader.mDeltaTimestamp.mValue = timeStampVal;
+        hasEpochTimestamp                   = true;
+    }
+
+    if ((hasSystemTimestamp && !hasEpochTimestamp && !hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && hasEpochTimestamp && !hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && !hasEpochTimestamp && hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && !hasEpochTimestamp && !hasDeltaSystemTimestamp && hasDeltaEpochTimestamp))
+    {
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_IM_MALFORMED_EVENT_DATA_ELEMENT;
+}
+
+CHIP_ERROR ReadClient::ProcessEventReportIBs(TLV::TLVReader & aEventReportIBsReader)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    while (CHIP_NO_ERROR == (err = aEventReportIBsReader.Next()))
+    {
+        TLV::TLVReader dataReader;
+        EventReportIB::Parser report;
+        EventDataIB::Parser data;
+        EventStatusIB::Parser status;
+        EventPathIB::Parser path;
+        ClusterInfo clusterInfo;
+        StatusIB statusIB;
+        EventHeader header;
+        uint8_t priorityLevel = 0;
+
+        TLV::TLVReader reader = aEventReportIBsReader;
+        ReturnErrorOnFailure(report.Init(reader));
+        // EventStatus and EventData would coexist
+        err = report.GetEventStatus(&status);
+        ReturnErrorOnFailure(err);
+        StatusIB::Parser errorStatus;
+        ReturnErrorOnFailure(status.GetPath(&path));
+        ReturnErrorOnFailure(ProcessEventPath(path, clusterInfo));
+        ReturnErrorOnFailure(status.GetErrorStatus(&errorStatus));
+        ReturnErrorOnFailure(errorStatus.DecodeStatusIB(statusIB));
+
+        ReturnErrorOnFailure(report.GetEventData(&data));
+        ReturnErrorOnFailure(data.GetPath(&path));
+        ReturnErrorOnFailure(ProcessEventPath(path, clusterInfo));
```

Comment:
Why do this intermediate conversion to clusterInfo, only to then convert to a ConcreteEventPath?

Reply:
Inside ProcessEventPath, we can reuse  HasWildcardEventId and check if it is IsValidEventPath in ClusterInfo, where it also has similar check for valid attributePath.

Reply:
resolved.